### PR TITLE
added @Category(SlowTest.class) and a gradle task to exclude them

### PR DESCRIPTION
**Pull request by migwellian**
A new gradle task that you can optionally run on your local machine if you want to exclude the slow tests. The 968 tests now run in about 10 seconds :-)

```
$  ./gradlew keanu-project:testFastOnes
BUILD SUCCESSFUL in 30s
12 actionable tasks: 3 executed, 9 up-to-date
```

### CODE REVIEW

File: keanu-project/build.gradle [hunk 1]
```diff
@@ -52,6 +52,12 @@ test {
     systemProperty "dtype", 'double'
 }
 
+task testFastOnes(type: Test) {
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/build.gradle]: I did... my thinking is that this should be the "power users" option. If it's the default then people are more likely to never run the slow tests, and therefore are in danger of checking in code that breaks the slow tests.

[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/build.gradle]: Perhaps a compromise would be if `test` only runs the fast ones, but `build` runs all the tests?

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/HamiltonianTest.java [hunk 2]
```diff
@@ -64,6 +67,7 @@ public void samplesContinuousPrior() {
         MCMCTestDistributions.samplesMatchesSumOfGaussians(44.0, posteriorSamples.get(A).asList(), posteriorSamples.get(B).asList());
     }
 
+    @Category(Slow.class)
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/HamiltonianTest.java]: I ran the tests in IntelliJ, which tells you the duration of each test. By and large I removed anything slower than 100ms, however I used some discretion, such as:
* don't mark every test in a class as `Slow`
* for consistency, I marked all of the `SampleMethodMatchesLogProbMethod`s as slow, even though some of them were <50ms

File: keanu-project/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/build.gradle]: Yeah, like a test vs check thing?  That would work for me, certainly.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/particlefiltering/ParticleFilteringTest.java [hunk 3]
```diff
@@ -18,6 +20,7 @@
 
     private final Logger log = LoggerFactory.getLogger(ParticleFilteringTest.class);
 
+    @Category(Slow.class)
```
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/particlefiltering/ParticleFilteringTest.java]: It's true, this contradicts my comment above about not marking every test in a class as `Slow`. Although there's only one test in this class - perhaps we can add some more?

Also don't think of this as "we're not running these tests". The idea is that the user should always do a full build and test before pushing to github.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/HamiltonianTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/HamiltonianTest.java]: Makes sense.  The Donut test seems a bit more interesting to me than the continuous one I guess, but not a big deal.

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/particlefiltering/ParticleFilteringTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (christophernorth) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/particlefiltering/ParticleFilteringTest.java]: Oh - yeah, and obviously CI will do a full run too - I'm just thinking of the Psychology of engineers - they're likely to save a full run until the last minute - and we all know the later a bug is found, often the more costly it is to fix.  So yeah, I think we should keep one of these around personally...

File: keanu-project/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/build.gradle]: done

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/HamiltonianTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/mcmc/HamiltonianTest.java]: I'm gonna leave it as it is - on my machine the donut takes up to twice as long as the continuous (1s vs 0.5s)

File: keanu-project/src/test/java/io/improbable/keanu/algorithms/particlefiltering/ParticleFilteringTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (migwellian) [reply keanu-project/src/test/java/io/improbable/keanu/algorithms/particlefiltering/ParticleFilteringTest.java]: I'll stick it back in. It takes about a third of a second to run.

**Pull request merged**